Bounding Box Camera

AmrElsersy opened the pull request:
## 🎉 New feature

related to [#135](https://github.com/ignitionrobotics/ign-sensors/issues/135)

## Summary
Bounding boxes camera to extract labeled item's bounding boxes from the scene using OGRE 2.1

## Visible Bounding Boxes
- Supports visible bounding boxes which shows a box on the visible part of the item (box dosen't include the occluded part of the object) (datasets ex: COCO & PASCAL VOC datasets) 
- Idea: Rendering a map which contains ogre Ids of objects, and by looping on its pixels to get the the objects boundaries 

<img src="https://user-images.githubusercontent.com/35613645/121053995-37454e00-c7bc-11eb-809a-5fd556494039.gif" width="640" height="480" />


## Full Bounding Boxes
- Supports full bbox which shows all the object boundaries even if occluded (datasets ex: KITTI, nuscenes)
- Idea: accessing all item's vertices and projecting them to get the boundaries of the object, and check visibility by rendering a map which contains ogre Ids of objects

<img src="https://user-images.githubusercontent.com/35613645/121054371-91deaa00-c7bc-11eb-8040-8de7a7eec221.gif" width="640" height="480" />


## Features:
- The user label each item by setting its UserData
- Each unlabeled item is considered as a background (not finished yet)
- Added example in ign-rendering/examples to try it easily
- User can change the mode (Visible / Full) through the rendering API

AmrElsersy commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,824 @@
+      {
+        // items with no class are considered background
+        label = this->backgroundLabel;
+      }
```
There is a problem in reading UserData of the visual, when the visual has no user data of the label key ( "label" ) .. the `std::get<int>(labelKey)` returns 0 and dosn't through the bad_variant_access exception
Any idea about why is that ?

AmrElsersy commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,824 @@
+      int label = this->backgroundLabel;
+      try
+      {
+        label = std::get<int>(labelAny);
```
There is a problem in reading UserData of the visual, when the visual has no user data of the label key ( "label" ) .. the std::get<int>(labelKey) returns 0 and dosn't through the bad_variant_access exception
Any idea about why is that ?

AmrElsersy replied:
@adlarkin

adlarkin replied:
Looking at the code, I don't see anything strange that sticks out to me. Do you have an example setup I can try that replicates the issue? Maybe an SDF file or something like that

adlarkin commented:
@AmrElsersy here are some ideas for integration testing:
1. Generate a scene with 2 objects: one that is completely to the left of another. Make sure that you end up with 2 bounding boxes, and that all of the coordinates for the left object's bounding box are actually to the left of or right object's bounding box (so, this would probably involve checking that all of the x coordinates for one object are less than all of the x coordinates for the other object)
2. Test objects that overlap one another:
    1. two objects, with one object completely behind the other. We should only get one bounding box in this case, and the bounding box should be for the object that is not occluded
    2. two objects, with one object partially behind the other. We can then test the visible and full bounding boxes - the partially blocked object's bounding box for the visible bounding box sensor should be smaller than the partially blocked object's bounding box for the full bounding box sensor

Also, a note about checking the validity of bounding box coordinates: you can see if a bounding box's corner pixel values are _close enough_ to a "baseline" pixel value by using [`EXPECT_NEAR`](https://github.com/google/googletest/blob/53495a2a7d6ba7e0691a7f3602e9a5324bba6e45/docs/reference/assertions.md#expect_near-expect_near)

chapulina commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,90 @@
+
+    /// \brief Bounding box structure to store its boundaries
+    /// min & max of x & y, and the label
+    struct BoundingBox
```
You could use `ignition::math::AxisAlignedBox` instead of creating a new type. That's meant for 3D boxes, but I think it should be ok to leave the Z dimension as zero when doing 2D :thinking:

AmrElsersy replied:
@chapulina 
It only provides the bounding boxes, that RGB image because in the example I have 2 cameras, and I draw the bboxes on the image of the rgb camera (just for seek of visualization and trying the example)

mmmm, yeah I have a plan to see how to provide the 3d boxes with the lidar data, but that after I finish all thing in the project :D .. do you think we should support 3d boxes with images instead of lidar ?

about the `math::AxisAlignedBox ` .. I created a new struct because I had to store the label of the bounding box with the dimensions

chapulina replied:
> It only provides the bounding boxes, that RGB image because in the example I have 2 cameras,

Ah perfect!

> do you think we should support 3d boxes with images instead of lidar ?

We have access to the ground truth from simulation, so I think we could just get the 3D bounding box of the visuals that are within a given volume (like a camera frustum?). If we want to take occlusion into account, I think we need to define the observation point, like a camera's image plane.

> I created a new struct because I had to store the label of the bounding box with the dimensions

Yeah I noticed that. You could use the `AxisAlignedBox` inside your struct. The advantage is that the class offers various helper functions that could be useful for users.

AmrElsersy replied:
That is because this PR is done before doing the msgs PR
But it is much simpler to use the struct instead of the msg, as i pass this class everywhere in the code
can we just keep it ?

chapulina reviewed (commented):
The example looks super cool! 

Can you comment on what's the intended use case? I see the sensor outputs the bounding box coordinates, as well as an RGB image with the boxes drawn. I'm wondering if we need the RGB image too, or if this could be a simpler sensor that just provides the BBs. That would make the sensor more similar to `GpuRays` than to a camera.

Another thing that came to mind is that this camera is providing the 2D bounding boxes, but 3D bounding boxes are also widely used in ML datasets. Are there plans to add 3D boxes too?

AmrElsersy commented:
I am trying to add the 3D boxes feature but have a problem in the oreintation of the bbox
That is because the OGRE item has a Quaternion that specifies the rotation around the object axes.
and we have the view matrix which specifies the rotation of the camera 

I want to add these 2 Quaternions 
Does anyone know how to do that ?, I am trying to add the euler angles of both of them , but it is not working, 

I think that is because the Item Quaternion is in World coord, so roll, pitch and yaw is the rotation around x, y and z ... but the view matrix Quaternion is in camera coord, so roll, pitch, yaw are rotations around different axes than the world coord 

Any ideas on how to do that ?

This photo shows that the 3d position is correct but has a wrong rotation
![Screenshot 2021-07-14 11:44:13](https://user-images.githubusercontent.com/35613645/125602123-bd992ff5-a328-4bac-a0b2-69f56b060629.png)

mjcarroll commented:
It sounds like you have two rotations:

1. The rotation from body to world `q_b_w`
2. The rotation from camera to world. `q_c_w`

You should be able to multiply these two quaternions to get the rotation from body to camera, but you may need to invert one first.

So `q_c_b = q_c_w * inv(q_b_w)`

iche033 commented:
by the way, the [ViewMatrix()](https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering5/include/ignition/rendering/base/BaseCamera.hh#L508) call does additional conversion from y to z up. Not sure if you need that. Otherwise, you can just call `camera->WorldRotation()` to get the camera rotation in world frame

AmrElsersy commented:
> It sounds like you have two rotations:
> 
> 1. The rotation from body to world `q_b_w`
> 2. The rotation from camera to world. `q_c_w`

is the rotation from camera to world (q_c_w), the quaternion from the view matrix ??

AmrElsersy commented:
I tried what you mention @mjcarroll  , @iche033 but I can't make it work 
...
I want to get the rotation from body coord to camera coord, and what i have is 
1- the body to world quaternion, from `ogreNode->getOrientation( )`
2- the world to camera quaternion from `camera->ViewMatrix().extractQuaternion()`
3- the world to camera rotation from `camera->WorldRotation()`

I tried to do
`bodyCameraRotation = viewMatrixQuaternion * bodyWorldRotation`

I tried to inverse each one of them also
tried to replace the viewMatrixQuaternion with the camera->WorldRotation()
but also not working

AmrElsersy commented:
It is working now, thanks :D @iche033  @mjcarroll 
![111111](https://user-images.githubusercontent.com/35613645/125708638-e830b3f6-9312-42ef-9983-1a5a317023d0.png)

AmrElsersy commented:
It seems like it is a problem of the way we get the data from the variant 
the UserData(key) of the visual returns a Variant object with no value if the key is not found
when we access the data via
```cpp
  auto x = std::get<int>(variant); 
  // x = 0
```
will return 0 always (will through exception in case of <double> <float> <string>
So we need to add a way to specify an empty variant instead of returning a variant with no value setted yet

I tried to modify the Variant / UserData .. and it is working (and passes the test)
Visual.hh
```cpp
    using Variant = std::variant<int, float, double, std::string, std::monostate>;
```
BaseVisual.hh
```cpp
    template <class T>
    Variant BaseVisual<T>::UserData(const std::string &_key) const
    {
      auto it = this->userData.find(_key);
      if (it != this->userData.end())
        return it->second;
      else
        return std::monostate();
    }
```

@adlarkin

AmrElsersy commented:
Solve the problem of [Multi Link Models](https://github.com/ignitionrobotics/ign-gazebo/pull/853#issuecomment-882320969) for 2D bounding boxes by merging all the boxes that belong to the same parent

![boxes_multi_model](https://user-images.githubusercontent.com/35613645/126617825-9a727ee5-ce94-4d0c-98b6-63945bc94ee8.png)


Anyone have an idea for how to merge the 3D oriented boxes ?? because each part will have a different rotation in all directions, so i don't know yet how to do it

AmrElsersy commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,112 @@
+  namespace rendering
+  {
+    /// \brief BoundingBox types for Visible / Full BoundingBox
+    enum class BoundingBoxType {
+
+      /// FullBox2D: 2D box that shows the full box of occluded objects
```
It will not be readable here if it becomes uppercase :D 
I see that notation is also used .. https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/include/ignition/gazebo/Types.hh#L63

can we keep that also :D

iche033 replied:
let's make these uppercase for consistency within ign-rendering. It should also have a prefix, i.e. `BBT_FULLBOX2D`

AmrElsersy commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,112 @@
+
+      /// \brief Get the BoundingBox data
+      /// \return Buffer of bounding boxes info (label, minX, minY, maxX, maxY)
+      public: virtual std::vector<BoundingBox> BoundingBoxData() const = 0;
```
I need to clear the bounding boxes data inside the class, so i have to return by copy

AmrElsersy commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1078 @@
+
+    // get the derived position
+    Ogre::Vector3 position = node->_getDerivedPositionUpdated();
+    Ogre::Quaternion oreintation = node->_getDerivedOrientationUpdated();
```
is there any problem with using the internal functions of OGRE ??

adlarkin replied:
I'm not sure, @iche033 what are your thoughts? I would think that we should try to shy away from internal OGRE methods (i.e., ones that start with `_`), but it does seem like they're being used in a few other places in `ign-rendering`.

iche033 replied:
these should be fine. If possible, use the `_getDerivedPosition` (without `Updated`) version to avoid extra node updates.

adlarkin commented:
> It seems like it is a problem of the way we get the data from the variant
> the UserData(key) of the visual returns a Variant object with no value if the key is not found
> when we access the data via
> 
> ```c++
>   auto x = std::get<int>(variant); 
>   // x = 0
> ```
> 
> will return 0 always (will through exception in case of
> So we need to add a way to specify an empty variant instead of returning a variant with no value setted yet
> 
> I tried to modify the Variant / UserData .. and it is working (and passes the test)
> Visual.hh
> 
> ```c++
>     using Variant = std::variant<int, float, double, std::string, std::monostate>;
> ```
> 
> BaseVisual.hh
> 
> ```c++
>     template <class T>
>     Variant BaseVisual<T>::UserData(const std::string &_key) const
>     {
>       auto it = this->userData.find(_key);
>       if (it != this->userData.end())
>         return it->second;
>       else
>         return std::monostate();
>     }
> ```
> 
> @adlarkin
> 
> I noticed that the implementation of the user data is transfered to Node instead of Visual
> I merged that and modified it again, anyway, should I put these changes as a commit or in a separate PR ?

It turns out that this is actually a behavioral bug in `ign-rendering`, which means it needs to be fixed in all versions of `ign-rendering`. I have opened a PR to fix this in `ign-rendering3` (see #370), and will port it forward to the `main` branch once it is merged. Once the fix is ported forward to `main`, all you'll need to do is update your branch with the latest changes from `main` and then this behavior should be resolved. Once you have these changes merged into your PR, the change you made above will no longer be needed.

adlarkin commented:
> Anyone have an idea for how to merge the 3D oriented boxes ?? because each part will have a different rotation in all directions, so i don't know yet how to do it

I would think that we want to go off of the rotation of the canonical link, since that's the "reference point" a user defines. Does anyone else have thoughts/opinions about this?

adlarkin commented:
@AmrElsersy, https://github.com/ignitionrobotics/ign-rendering/pull/371 has been merged, so if you update your branch with `main` and use the changes from #371, you should no longer have the issue with getting an int with a value of `0` for keys that don't exist with UserData.

azeey commented:
> Anyone have an idea for how to merge the 3D oriented boxes ?? because each part will have a different rotation in all directions, so i don't know yet how to do it

Looks like this can be a tough problem to solve accurately, but approximate solutions might be sufficient. Open3d has a function `OrientedBoundingBox::CreateFromPoints` (https://github.com/isl-org/Open3D/blob/76c2baf9debd460900f056a9b51e9a80de9c0e64/cpp/open3d/geometry/BoundingVolume.h#L123-L129). Can you just use the vertices of the bounding boxes of the links as the input to such an algorithm?

scpeters commented:
> > Anyone have an idea for how to merge the 3D oriented boxes ?? because each part will have a different rotation in all directions, so i don't know yet how to do it
> 
> Looks like this can be a tough problem to solve accurately, but approximate solutions might be sufficient. Open3d has a function `OrientedBoundingBox::CreateFromPoints` (https://github.com/isl-org/Open3D/blob/76c2baf9debd460900f056a9b51e9a80de9c0e64/cpp/open3d/geometry/BoundingVolume.h#L123-L129). Can you just use the vertices of the bounding boxes of the links as the input to such an algorithm?

this sounds right to me. You can account for the difference in orientation between the bounding boxes by passing the vertices of each child box to the parent box and expand the parent box until it encompasses all the child vertices

AmrElsersy commented:
@azeey 
@scpeters 
Do you mean I have to add the `Open3D` library as a dependency to ignition ?? or to implement a similar algorithm ?? because it uses many local classes inside that function

azeey commented:
I was thinking we implement a similar algorithm in ign-math. I didn't look at how involved that would be.

AmrElsersy commented:
Ok will try to do that

adlarkin commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,109 @@
+      /// \brief Size of the box (width, height, depth), depth = 0 in 2D boxes
+      ignition::math::Vector3d size;
+
+      /// \brief Oreintation of the 3D box in camera coord.
+      /// The 2D boxes are axis aligned (oreintation = 0)
+      ignition::math::Quaterniond oreintation;
```
```suggestion
      /// \brief Orientation of the 3D box in camera coord.
      /// The 2D boxes are axis aligned (oreintation = 0)
      ignition::math::Quaterniond orientation;
```

AmrElsersy commented:
I am trying some open source approach for finding the oriented box from the 3d vertices of the mesh(point cloud) 
http://codextechnicanum.blogspot.com/2015/04/find-minimum-oriented-bounding-box-of.html

I tried it outside ignition, on some LIDAR's pointclouds and it works .. I am trying to integrate it now into ignition
It is based on Eigen library .. but I don't know where to put the code inside `ign-math` .. It seems that `ign-math` has no `Eigen` code except for the `Conversions.hh` file, I put the code in that file but i cannot include that file from `ign-rendering` to be able to use the code as Eigen library is not a dependency of ign-rendering.

That is a part of the code 
```cpp
      inline ignition::math::OrientedBoxd convert(const std::vector<math::Vector3d> &_mesh)
      {
        math::OrientedBoxd box;

        // Centroid (mean) of the mesh
        math::Vector3d mean;
        for (auto point : _mesh)
          mean += point;
        mean /= _mesh.size();

        Eigen::Vector3d centroid = convert(mean);

        Eigen::Matrix3d covariance = computeCovarianceMatrixNormalized(_mesh, centroid);
        Eigen::SelfAdjointEigenSolver<Eigen::Matrix3d> eigenSolver(covariance, Eigen::ComputeEigenvectors);
        Eigen::Matrix3d eigenVectorsPCA = eigenSolver.eigenvectors();

        eigenVectorsPCA.col(2) = eigenVectorsPCA.col(0).cross(eigenVectorsPCA.col(1));

        Eigen::Matrix4d projectionTransform(Eigen::Matrix4d::Identity());
        projectionTransform.block<3,3> (0,0) = eigenVectorsPCA.transpose();
        projectionTransform.block<3,1> (0,3) = -1.f * (projectionTransform.block<3,3>(0,0) * centroid.head<3>());
        ...
```

So for short, where can i put a code which is based on Eigen library, and use it from ign-rendering ?

@azeey 
@adlarkin 
@scpeters

AmrElsersy commented:
And btw, it is the same algorithm that is used in `Open3D` .. and I will try to use vertices of the boxes instead of the whole vertices of the mesh to see if it is acceptable 

But anyway i still have the same problem of where to put the Eigen code :D

adlarkin commented:
> Eigen library is not a dependency of ign-rendering

`ign-rendering` depends on `ign-math`, and `ign-math` depends on `Eigen`. So, if you place the relevant code in `ign-math` and then use the `ign-math` APIs in `ign-rendering`, you shouldn't have any problems _as long as you make `ign-rendering` depend on the `eigen3` component of `ign-math` (see more information about how to do this below)._

> where can i put a code which is based on Eigen library

All code that uses `Eigen` in `ign-math` should be placed in [`ign-math`'s `eigen3` folder](https://github.com/ignitionrobotics/ign-math/tree/ignition-math6_6.8.0/eigen3/include/ignition/math/eigen3). Currently, the only file in that folder is `Conversions.hh`, which is meant to provide conversions between `Eigen` and `ign-math` types. Since your method takes an `ign-math` type as input and returns and `ign-math` type (it just uses `Eigen` to help with the implementation), I don't think it should be placed in `Conversions.hh`. We should probably create a new file in the `eigen3` folder for this code. However, I'm not sure what @azeey and @scpeters think about this, so for now, place it in `Conversions.hh` and then we can move it to a new file if needed.

> how to use it from ign-rendering ?

The eigen code in `ign-math` is a separate component: https://github.com/ignitionrobotics/ign-math/blob/ignition-math6_6.8.0/eigen3/include/ignition/math/CMakeLists.txt

So, I believe that you'll need to specify this component when searching for `ign-math` in `ign-rendering`'s `CMakeLists.txt`. Take a look at `ign-gazebo` as an example:
* depending on the `ign-math` `eigen3` component in the `CMakeLists.txt`: https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.1.0/CMakeLists.txt#L122
* Including this header and using it in a `ign-gazebo` file: https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.1.0/src/systems/physics/Physics.cc#L38 and https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.1.0/src/systems/physics/Physics.cc#L1028

AmrElsersy commented:
I still cannot include the file (`Conversions.hh`) 
I added the eigen component in `ign-rendering/CMakeLists.txt`
```
ign_find_package(ignition-math6 REQUIRED COMPONENTS eigen3 VERSION 6.8)
```
instead of 
```
# ign_find_package(ignition-math6 REQUIRED VERSION 6.6)
```

but it is not working, I can not include the `#include <ignition/math/eigen3/Conversions.hh>`
it gives me 
```
cannot open source file "Eigen/Geometry" (dependency of "ignition/math/eigen3/Conversions.hh"
```

@adlarkin

scpeters commented:
> but it is not working, I can not include the `#include <ignition/math/eigen3/Conversions.hh>`
> it gives me
> 
> ```
> cannot open source file "Eigen/Geometry" (dependency of "ignition/math/eigen3/Conversions.hh"
> ```

did you add the eigen component to the `target_link_libraries`?

https://github.com/ignitionrobotics/ign-physics/blob/f38120bcde5a019eaab9e6342ae0356c50243397/dartsim/CMakeLists.txt#L35

AmrElsersy commented:
> did you add the eigen component to the `target_link_libraries`?
> 
> https://github.com/ignitionrobotics/ign-physics/blob/f38120bcde5a019eaab9e6342ae0356c50243397/dartsim/CMakeLists.txt#L35

Thanks Steve it worked :D

azeey commented:
> We should probably create a new file in the eigen3 folder for this code. However, I'm not sure what @azeey and @scpeters think about this, so for now, place it in Conversions.hh and then we can move it to a new file if needed.

+1 for adding a new file. I believe the code in `Conversions.hh` is for converting between equivalent types of ign-math and Eigen while this is introducing a new algorithm for creating a bounding box from a collection of points.

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,190 @@
+*/
+
+#ifndef IGNITION_RENDERING_OGRE2_OGRE2BoundingBoxCAMERA_HH_
+#define IGNITION_RENDERING_OGRE2_OGRE2BoundingBoxCAMERA_HH_
```
make this all uppercase

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,190 @@
+      public: virtual void MergeMultiLinksModels3D();
+
+      /// \brief Pointer to the ogre camera
+      protected: Ogre::Camera *ogreCamera;
```
set to `nullptr`

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,190 @@
+      /// to the same parent (only used in multi-links models)
+      /// \param[in] _ogreIds vector of ogre ids that belongs to the same model
+      /// \param[out] _vertices vector of 3d vertices of the item
+      public: virtual void GetMeshVertices(
```
style nitpick: remove `Get` prefix

iche033 commented on test/integration/boundingbox_camera.cc:
```diff
@@ -0,0 +1,459 @@
+
+  public: void SimpleBoxes(const std::string &_renderEngine);
+
+  public: void Oreinted3dBoxes(const std::string &_renderEngine);
```
`Oriented3dBoxes`

iche033 commented on test/integration/boundingbox_camera.cc:
```diff
@@ -0,0 +1,459 @@
+class BoundingBoxCameraTest: public testing::Test,
+  public testing::WithParamInterface<const char *>
+{
+  public: void OccludedBoxes(const std::string &_renderEngine);
```
can you add some brief comments to describe the tests?

iche033 commented on test/integration/boundingbox_camera.cc:
```diff
@@ -0,0 +1,459 @@
+  root->AddChild(invisibleBox);
+}
+
+/// \brief Build a scene with a single oreinted box for 3D test
```
`oreinted` -> `oriented`

iche033 commented on examples/boundingbox_camera/GlutWindow.cc:
```diff
@@ -0,0 +1,450 @@
+    auto yaw = std::to_string(box.orientation.Yaw());
+
+    std::string sep = " ";
+    std::string boxString = x + sep + y + sep + z + sep + w + sep + h + sep + l + sep +
```
the `math::Vector3d` and `math::Quaterniond` implements the `<<` operator so you should be able to just do ` << box.center << " " << box.size << " " << box.orientation`

iche033 commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,109 @@
+      /// \return BoundingBox Type (Visible / Full)
+      public: virtual BoundingBoxType Type() const = 0;
+
+      /// \brief Draw a bounding box on the givin image in green color
```
given

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,190 @@
+          const Ogre::Vector3 &scale
+          );
+
+      /// \brief Draw bounding box on the givin image
```
`givin` -> `given`

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,190 @@
+      public: virtual void MergeMultiLinksModels2D();
+
+      /// \brief Merge a links's 3d boxes of multi links models
+      public: virtual void MergeMultiLinksModels3D();
```
can we make functions that are not used outside this class `private`? Same applies to the `OgreBoundingBoxCamera` class

iche033 commented on examples/boundingbox_camera/GlutWindow.cc:
```diff
@@ -0,0 +1,450 @@
+{
+  std::cout << "===============================" << std::endl;
+  std::cout << "  TAB - Switch render engines  " << std::endl;
+  std::cout << "  ESC - Exit                   " << std::endl;
```
can you add an entry for `s`?

iche033 commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,109 @@
+ *
+ */
+#ifndef IGNITION_RENDERING_BoundingBOXCAMERA_HH_
+#define IGNITION_RENDERING_BoundingBOXCAMERA_HH_
```
change this to all to uppercase

iche033 commented on examples/boundingbox_camera/GlutWindow.hh:
```diff
@@ -0,0 +1,29 @@
+ * limitations under the License.
+ *
+ */
+#ifndef IGNITION_RENDERING_EXAMPLES_SEGMENTATION_CAMERA_GLUTWINDOW_HH_
```
`SEGMENTATION` -> `BOUNDINGBOX`

iche033 commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1230 @@
+#include "ignition/rendering/ogre2/Ogre2Includes.hh"
+#include "ignition/rendering/ogre2/Ogre2RenderTarget.hh"
+#include "ignition/rendering/ogre2/Ogre2Scene.hh"
+#include "ignition/rendering/ogre2/Ogre2SelectionBuffer.hh"
```
is selection buffer needed?

AmrElsersy replied:
no

iche033 commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1230 @@
+#include <OgreGL3PlusAsyncTicket.h>
+#include <OgreBitwise.h>
+#include <OgreWireAabb.h>
+#include <OgreWireBoundingBox.h>
```
I think only the `OgreBitwise.h` is needed here and others can be removed?

iche033 commented on examples/boundingbox_camera/GlutWindow.cc:
```diff
@@ -0,0 +1,450 @@
+  g_camera->Capture(*g_image);
+}
+
+void initBoundingBoxCamera(ir::CameraPtr _camera)
```
add separator line:

```
//////////////////////////////////////////////////
```

iche033 reviewed (commented):
Did a first pass review. I tried the bounding box example but I don't see the red boxes. Is there anything I need to do for it to draw those bounding boxes?

AmrElsersy commented:
I added the feature of merging the 3d oriented bounding boxes ... it uses the code in ign-math in that PR [math#224](https://github.com/ignitionrobotics/ign-math/pull/224)

the idea is to get the 3d vertices of all childern links and then getting the oriented box from that vertices using PCA ... I used an open source code for that http://codextechnicanum.blogspot.com/2015/04/find-minimum-oriented-bounding-box-of.html 
and integrated it into ign-math and implemented some of the functions that the code use(as it uses pcl library in some cases)

The result is not perfect, but acceptable i think :D 

![merge3d](https://user-images.githubusercontent.com/35613645/129962049-20a928b1-ccf7-4a50-a50f-4fb824708ccb.png)
 
@azeey @scpeters  @adlarkin

AmrElsersy commented:
> Did a first pass review. I tried the bounding box example but I don't see the red boxes. Is there anything I need to do for it to draw those bounding boxes?

Thanks Ian :D 
Yeah that is because I was testing 3D boxes in the example (we don't have visualization for 3d boxes as it requires drawing a line with any angle, and i don't know how to do that), i edited it now to be 2d boxes, you can try it now (you probably need to get the ign-math PR code as the PR depends on it now https://github.com/ignitionrobotics/ign-math/pull/224

iche033 commented on examples/boundingbox_camera/GlutWindow.cc:
```diff
@@ -0,0 +1,436 @@
+#define KEY_TAB  9
+#define UNSUPPORTED_BUTTONS 5
+#define NORMALCAMERA_INDEX 0
+#define BOUNDINGBOX_INDEX 1
```
`NORMALCAMERA_INDEX` and `BOUNDINGBOX_INDEX` are not used and can be removed?

iche033 commented on include/ignition/rendering/BoundingBoxCamera.hh:
```diff
@@ -0,0 +1,109 @@
+      /// occluded object
+      BBT_VISIBLEBOX2D = 1,
+
+      /// 3D oreinted box
```
`oreinted` -> `oriented`

iche033 commented on CMakeLists.txt:
```diff
@@ -29,7 +29,8 @@ ign_configure_project(VERSION_SUFFIX pre1)
 #--------------------------------------
 # Find ignition-math
-ign_find_package(ignition-math6 REQUIRED VERSION 6.6)
+# ign_find_package(ignition-math6 REQUIRED VERSION 6.6)
```
remove?

iche033 reviewed (commented):
> we don't have visualization for 3d boxes as it requires drawing a line between any 2 points in the image, and i don't know how to do that

looks like you're drawing the boxes by directly changing the pixel color in the image. So I think what you need is a way to get 2d image positions of the 3d points of the bounding box corners? If so, you can use the `Camera::Project` function:
https://github.com/ignitionrobotics/ign-rendering/blob/main/include/ignition/rendering/Camera.hh#L228

AmrElsersy commented:
> looks like you're drawing the boxes by directly changing the pixel color in the image. So I think what you need is a way to get 2d image positions of the 3d points of the bounding box corners? If so, you can use the `Camera::Project` function:

Yeah, I got that already, I have all the 2D points to draw the 3D box on the image
But I don't know how to draw a line between any two 2D points in the image buffer, I only know how to draw vertical or horizontal lines @iche033

osrf-triage commented:
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/gsoc-2021-machine-learning-extension-to-ignition-gazebo/1070/1

iche033 commented:
> But I don't know how to draw a line between any two 2D points in the image buffer, I only know how to draw vertical or horizontal lines @iche033

I see. Is this what you are looking for?
https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm

AmrElsersy commented:
Thanks Ian for mentioning the algorithm :D, we now have visualization for the 3D boxes 

![3d_visualize](https://user-images.githubusercontent.com/35613645/131930029-12454c14-26c6-482e-a990-d86bcd79515c.png)

iche033 commented:
nice! 3D box is working well in most cases. I think there some some corner cases that don't look right when I zoom close in at an object:

![bounding_camera_3dbox](https://user-images.githubusercontent.com/4000684/132063573-b55e1992-8d19-4c3c-ba4d-7f59ca11e478.png)

AmrElsersy commented:
> nice! 3D box is working well in most cases. I think there some some corner cases that don't look right when I zoom close in at an object:
> 

@iche033 

Ok, I found why it goes like that, when we zoom in, the projected 2d points of the 3d vertices of the box, become outside the clip space [-1, 1], so we get screen coordinates outside the image size,

I clip these coordinates to the image size (ex: if the point has x coordinates > the width of the image, i clip it to the image width), that is why there are many points in the top right corner in your image.

.....
So what is the desired behavior for that ? or what should I do when I have 2d points that is outside the image size ?

iche033 commented:
> So what is the desired behavior for that ? or what should I do when I have 2d points that is outside the image size ?

ah ok I see.  How about instead of clipping the coordinates,  just draw the line until x or y is equal to 0 or image width/height?

adlarkin commented:
> How about instead of clipping the coordinates, just draw the line until x or y is equal to 0 or image width/height?

I tried to achieve this by implementing https://en.wikipedia.org/wiki/Cohen%E2%80%93Sutherland_algorithm in https://github.com/ignitionrobotics/ign-rendering/pull/334/commits/5467cd981e207a3618486c82c1bb587fce39f145. It seems to work in some cases, but there are still issues. So, I am trying to debug and figure out what's wrong.

![clipping_failing](https://user-images.githubusercontent.com/42042756/133833504-d4aa134a-c693-41da-bd81-c332399514c5.gif)

AmrElsersy commented:
> I tried to achieve this by implementing https://en.wikipedia.org/wiki/Cohen%E2%80%93Sutherland_algorithm in [5467cd9](https://github.com/ignitionrobotics/ign-rendering/commit/5467cd981e207a3618486c82c1bb587fce39f145). It seems to work in some cases, but there are still issues. So, I am trying to debug and figure out what's wrong.

I think there is a case that we should hide the box totally, when we zoom in and some of the 3d vertices of the box go behind the camera view .. in that case, the lines drawn from these hidden vertices should not be visible as they are behind the camera.

I think that what is happening here in the gif, when you zoom in to the object, the algorithm tries to draw a line which has 2 points behind the camera.

chapulina commented:
It looks like this may not make it into the initial Fortress release, which is fine. In order to get it into a minor release, we should open a separate PR with stubs of just the public API / ABI changes, printing warnings in case they're called.

AmrElsersy commented:
We skip boxes which has any vertex behind the camera and the visualization is more realistic now

![ezgif com-gif-maker (3)](https://user-images.githubusercontent.com/35613645/134267816-89a21f2d-6029-44bd-81da-ce3f91ba4e0d.gif)

chapulina commented:
> We skip boxes which has any vertex behind the camera and the visualization is more realistic now

Such a great idea! We can always add support for partial boxes later

chapulina commented:
Removing `beta` from this in favor of #420, which has the ABI / API breaking parts of this PR.

chapulina commented:
I'll move this to "In progress" while it hasn't been updated for the latest API.

adlarkin commented:
> I started working on fixing conflicts in [097a3e6](https://github.com/ignitionrobotics/ign-rendering/commit/097a3e6a340ee5380d63db49ab79a49f9d999e93), but I don't see any boxes when running the example, so I'm doing something wrong.

I looked at your changes and tried running the example from your branch. I also could not see any boxes, but I'm not sure why. Could it possibly be related to other changes that have occurred upstream on `ign-rendering6`? Maybe it would be best to convert your changes into a PR that targets this one so that we can debug/have a discussion there.

chapulina commented:
>  Maybe it would be best to convert your changes into a PR that targets this one so that we can debug/have a discussion there.

I didn't do that originally because I wondered if it would be best to start from scratch again. I'll give it another try and open a PR :+1:

adlarkin commented:
Can you elaborate a little more on what you mean by "start from scratch"?

chapulina commented:
> what you mean by "start from scratch"?

Ahh I meant just my merge attempt :smile:

adlarkin commented:
> Ahh I meant just my merge attempt

We can do that as well, and add a note that functionality is no longer working starting from that commit. We can always revert the commit if it seems too problematic. Merging here may also be best because the diff will be cleaner from the merge.

adlarkin commented:
@AmrElsersy, sorry for taking so long to get this merged. I'd like to revisit this and try to get it merged into Fortress soon. Since you no longer have an obligation to work in this feature since GSOC is over, what I will probably do in a week or two is cherry-pick your commits from this fork to a new branch in `ign-rendering` that targets `ign-rendering6`. That way, myself and other ignition maintainers can work on the final changes directly while still giving you credit for all of the commits you made. If you are against this approach in any way and would rather work on final iterations yourself on this fork, just let me know - otherwise, I'll move this over to a new PR in a week or two. Thanks!

adlarkin commented:
@chapulina I merged in upstream changes from `ign-rendering6` in https://github.com/ignitionrobotics/ign-rendering/pull/334/commits/950d32c61bcf020c923bf1fd46b7029d3bb4a956 and fixed the build to use the new changes in https://github.com/ignitionrobotics/ign-rendering/pull/334/commits/dd2a78131502d7a97dbc32009af83e4a48dd3c92. So, everything should be up to date now and ready for review. I tested the bounding box example and was able to confirm that things are working with the new changes :tada: here are some things that will need to be considered when reviewing the updated PR:
* I added some new methods to the [`BoundingBox` class](https://github.com/ignitionrobotics/ign-rendering/blob/dd2a78131502d7a97dbc32009af83e4a48dd3c92/include/ignition/rendering/BoundingBox.hh), but tests still need to be added for this class.
* There are compiler warnings about a few type casts from unsigned ints to doubles. These will need to be addressed, but I'm not sure if the fix is as simple as static casting from one type to the other (I think this will require taking a closer look at the code).
* For models composed of multiple 3D oriented bounding boxes, the object's merged 3D oriented bounding box is calculated via PCA (see https://github.com/ignitionrobotics/ign-rendering/pull/334#issuecomment-901382144). While this may be an acceptable solution _for now_, it seems odd to me that this should be a "ground truth" bounding box sensor that cannot produce ground truth merged 3D oriented bounding boxes. I think it would be worth re-visiting this and improving the implementation of computing merged 3D oriented boxes. If we want to stick with PCA for now so that this PR can be merged, it would be worth creating an issue so that this can be tracked and re-visited later.
* There are some edge cases where an object is visible, but no bounding box is present (see the image below; there is no bounding box for the blue box in the corner). I think this is related to https://github.com/ignitionrobotics/ign-rendering/pull/334#issuecomment-924501192. If we don't plan to complete partial box support in this PR, a ticket should be created so that this can also be tracked and re-visited later.

![missing_bbox](https://user-images.githubusercontent.com/42042756/164559692-6b7fa65e-92c9-43d9-932e-ee883ce775c0.png)

AmrElsersy commented:
> Changing the FOV seems to be fine, but hen changing the aspect ratio, the bounding boxes seem to have misalignment.
> 
 I will take a look at that 👍

AmrElsersy commented:
> Changing the FOV seems to be fine, but hen changing the aspect ratio, the bounding boxes seem to have misalignment.

I think you changed the aspect ratio of the camera but didn't change the width & height of the image according to that aspect ratio right ?

I tried to change the aspect ratio as well as a suitable width & height for it and it worked without any problems

WilliamLewww commented:
> > Changing the FOV seems to be fine, but hen changing the aspect ratio, the bounding boxes seem to have misalignment.
> 
> I think you changed the aspect ratio of the camera but didn't change the width & height of the image according to that aspect ratio right ?
> 
> I tried to change the aspect ratio as well as a suitable width & height for it and it worked without any problems

That fixed the issue for me, thanks!

WilliamLewww commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1469 @@
+
+  uint32_t width = maxX - minX;
+  uint32_t height = maxY - minY;
+  mergedBox.Size({width, height, 0});
+  mergedBox.Center({minX + width / 2, minY + height / 2, 0});
```
```suggestion
  mergedBox.Size({static_cast<double>(width), static_cast<double>(height), 0});
  
  mergedBox.Center({static_cast<double>(minX) + width / 2,
      static_cast<double>(minY) + height / 2, 0});
```

Prevent narrowing conversion warning.

WilliamLewww commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1469 @@
+    auto boxWidth = boundary->maxX - boundary->minX;
+    auto boxHeight = boundary->maxY - boundary->minY;
+
+    box.second->Center({boundary->minX + boxWidth / 2,
+        boundary->minY + boxHeight / 2, 0});
+    box.second->Size({boxWidth, boxHeight, 0});
```
```suggestion
    box.second->Center({static_cast<double>(boundary->minX) + boxWidth / 2,
        static_cast<double>(boundary->minY) + boxHeight / 2, 0});

    box.second->Size({static_cast<double>(boxWidth),
        static_cast<double>(boxHeight), 0});
```

Prevent narrowing conversion warning.

chapulina reviewed (commented):
This is working for me, except for the `2d_full` box type. On the example, it crashes after I orbit with the mouse for a bit. It also happens on `gz-sim`. I'm not being able to get an informative backtrace though.

AmrElsersy commented:
> This is working for me, except for the 2d_full box type. On the example, it crashes after I orbit with the mouse for a bit. It also happens on gz-sim. I'm not being able to get an informative backtrace though.

@chapulina  I will try to solve that :D

WilliamLewww commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1527 @@
+  std::vector<uint32_t> x_values =
+    {uint32_t(minVertex.X()), uint32_t(maxVertex.X())};
+  std::vector<uint32_t> y_values =
+    {uint32_t(minVertex.Y()), uint32_t(maxVertex.Y())};
```
Looks like the issue has to do with calculating the maxVertex. 

The example crashes when the bottom boundary line touches the bottom of the screen. Unlike the top, when the line reaches the bottom of the screen it either disappears or crashes the application.

Setting this line to the following code fixes the crash and also bounds the bottom line to the screen.
```
    {uint32_t(minVertex.Y()), uint32_t(maxVertexY() - 1)};
```

WilliamLewww replied:
```suggestion
    {uint32_t(minVertex.Y()),
        std::min(this->ImageHeight() - 1, uint32_t(maxVertex.Y()))};
```

WilliamLewww commented on ogre2/src/Ogre2BoundingBoxCamera.cc:
```diff
@@ -0,0 +1,1527 @@
+  uint32_t width = this->ImageWidth();
+
+  std::vector<uint32_t> x_values =
+    {uint32_t(minVertex.X()), uint32_t(maxVertex.X())};
```
```suggestion
    {uint32_t(minVertex.X()),
        std::min(this->ImageWidth() - 1, uint32_t(maxVertex.X()))};
```

chapulina replied:
Worked like a charm! Thanks!

WilliamLewww reviewed (changes requested):
Fix out of bounds error when creating lines for 2D_full

chapulina reviewed (approved):
All 3 camera types are working on the examples, tutorial, and downstream! Thanks to all the folks who got this across the finish line: @AmrElsersy @adlarkin @WilliamLewww ! 🏁
